src/content/docs/references/docs/defer.md

```diff
@@ -0,0 +1,161 @@
+---
+title: Defer and Cleanup
+description: Defer and Cleanup
+sidebar:
+    order: 115
+---
+
+# Defer 
+
+A `defer` *always* runs at the [end of a scope](#end-of-a-scope) at any point *after* it is declared, `defer` is commonly used to simplify code that needs clean-up; like closing unix file descriptors, freeing dynamically allocated memory or closing database connections.
+
+### End of a scope
+The end of a scope also includes `return`, `break`, `continue` or `!` rethrow. 
+
+[Rethrow](../optionals/#rethrow) `!` unwraps the optional, making it a normal variable again if [successful](../optionals) or if unsuccessful returns the [fault](../optionals) from the function back to the caller.
+
+```c3
+fn void test() 
+{
+    io::printn("print first");
+    defer io::printn("print third, on function return");
+    io::printn("print second");
+    return;
+}
+```
+The `defer` runs **after** the other print statments, at the function return.
+
+### Defer Execution order
+When there are multiple `defer` statements they are executed in reverse order of their declaration, last-to-first decalared. 
+
+
+```c3
+fn void test() 
+{
+    io::printn("print first");
+    defer io::printn("print third, defers execute in reverse order");
+    defer io::printn("print second, defers execute in reverse order");
+    return;
+}
+```
+
+### Example defer
+
+
+```c3
+import std::io;
+
+fn char[]! fileReader(String filename, char[] buffer)
+{   
+    io::File! file = file::open(filename, "r")!; // return if fault opening file
```

Comment:
Use `file_read` or something. `fileReader` is an unconventional name. Types are in general not prefixed with the module, so `File` rather than `io::File` here.